google_compute_engine_oslogin/utils/oslogin_utils.cc

```diff
@@ -493,4 +504,138 @@ bool ParseJsonToAuthorizeResponse(string response) {
   }
   return (bool)json_object_get_boolean(success);
 }
+
+bool ParseJsonToKey(const string& json, const string& key, string* response) {
+  json_object* root = NULL;
+  json_object* json_response = NULL;
+  const char* c_response;
+
+  root = json_tokener_parse(json.c_str());
+  if (root==NULL) {
+    return false;
+  }
+
+  if (!json_object_object_get_ex(root, key.c_str(), &json_response)) {
+    return false;
+  }
+
+  if (!(c_response = json_object_get_string(json_response))) {
+    return false;
+  }
+
+  *response = c_response;
+  return true;
+}
+
+bool ParseJsonToChallenges(const string& json, std::vector<Challenge> *challenges) {
+  json_object* root = NULL;
+
+  root = json_tokener_parse(json.c_str());
+  if (root == NULL) {
+    return false;
+  }
+
+  json_object *jsonChallenges = NULL;
+  if (!json_object_object_get_ex(root, "challenges", &jsonChallenges)) {
+    return false;
+  }
+
+  json_object *challengeId, *challengeType, *challengeStatus = NULL;
+  for (int i = 0; i < json_object_array_length(jsonChallenges); ++i) {
+    if (!json_object_object_get_ex(json_object_array_get_idx(jsonChallenges, i), "challengeId", &challengeId)) {
+      return false;
+    }
+    if (!json_object_object_get_ex(json_object_array_get_idx(jsonChallenges, i), "challengeType", &challengeType)) {
+      return false;
+    }
+    if (!json_object_object_get_ex(json_object_array_get_idx(jsonChallenges, i), "status", &challengeStatus)) {
+      return false;
+    }
+    Challenge challenge;
+    challenge.id = json_object_get_int(challengeId);
+    challenge.type = json_object_get_string(challengeType);
+    challenge.status = json_object_get_string(challengeStatus);
+
+    challenges->push_back(challenge);
+  }
+
+  return true;
+}
+
+bool GetUser(const string& username, string* response) {
+  std::stringstream url;
+  url << kMetadataServerUrl << "users?username=" << UrlEncode(username);
+
+  long http_code = 0;
+  if (!HttpGet(url.str(), response, &http_code) || response->empty() || http_code != 200) {
+    return false;
+  }
+  return true;
+}
+
+bool StartSession(const string& email, string* response) {
+  bool ret = true;
+  struct json_object *jobj, *jarr;
+
+  jarr = json_object_new_array();
+  json_object_array_add(jarr, json_object_new_string("INTERNAL_TWO_FACTOR"));
+  json_object_array_add(jarr, json_object_new_string("AUTHZEN"));
+  json_object_array_add(jarr, json_object_new_string("TOTP"));
+  json_object_array_add(jarr, json_object_new_string("IDV_PREREGISTERED_PHONE"));
```

Comment:
I might define these as constants to prevent typos both here and in the pam_module.